components/permalink_view/actions.js

```diff
@@ -3,18 +3,78 @@
 
 import {getChannel, getChannelMember, selectChannel, joinChannel, getChannelStats} from 'mattermost-redux/actions/channels';
 import {getPostThread} from 'mattermost-redux/actions/posts';
+import {getMissingProfilesByIds} from 'mattermost-redux/actions/users';
 import {getCurrentTeam, getTeam} from 'mattermost-redux/selectors/entities/teams';
 import {getCurrentUser, getUser} from 'mattermost-redux/selectors/entities/users';
+import {getCurrentChannel} from 'mattermost-redux/selectors/entities/channels';
+import {isCollapsedThreadsEnabled} from 'mattermost-redux/selectors/entities/preferences';
 
 import {loadChannelsForCurrentUser} from 'actions/channel_actions.jsx';
 import {loadNewDMIfNeeded, loadNewGMIfNeeded} from 'actions/user_actions.jsx';
+import {selectPostAndHighlight} from 'actions/views/rhs';
+
 import {browserHistory} from 'utils/browser_history';
 import {joinPrivateChannelPrompt} from 'utils/channel_utils';
 import {ActionTypes, Constants, ErrorPageTypes} from 'utils/constants';
 import {getUserIdFromChannelId, isSystemAdmin} from 'utils/utils';
+import {isComment} from 'utils/post_utils';
 
 let privateChannelJoinPromptVisible = false;
 
+function focusRootPost(post, team, channel) {
+    return async (dispatch, getState) => {
+        const postId = post.id;
+
+        dispatch(selectChannel(channel.id));
+        dispatch({
+            type: ActionTypes.RECEIVED_FOCUSED_POST,
+            data: postId,
+            channelId: channel.id,
+        });
+
+        if (channel.type === Constants.DM_CHANNEL) {
+            const userId = getUserIdFromChannelId(channel.name);
+            await dispatch(getMissingProfilesByIds([userId]));
+            const user = getUser(getState(), userId);
+            browserHistory.replace(`/${team.name}/messages/@${user.username}/${postId}`);
+        } else if (channel.type === Constants.GM_CHANNEL) {
+            browserHistory.replace(`/${team.name}/messages/${channel.name}/${postId}`);
+        } else {
+            browserHistory.replace(`/${team.name}/channels/${channel.name}/${postId}`);
+        }
+    };
+}
+
+function focustReplyPost(post, team, channel, returnTo) {
+    return async (dispatch, getState) => {
+        await dispatch(getPostThread(post.parent_id));
+
+        const state = getState();
+
+        const currentChannel = getCurrentChannel(state);
+        const sameTeam = currentChannel && currentChannel.team_id === team.id;
+
+        if (!sameTeam) {
+            dispatch(selectChannel(channel.id));
```

Comment:
Would we possibly need to switch teams here as well?

Reply:
the permalink contains the team, so that happens regardless.